blogs/posts/2024-02-11_sankey_plot.qmd

```diff
@@ -0,0 +1,565 @@
+    summarise(
+      flow = n_distinct(participant_id, na.rm = T),
+      .by = c(from, to)
+    ),
+  
+  # flow from invite 1 to invite 2 (where not NA)
+  df_milestones |> 
+    filter(!is.na(invite_1_outcome) & !is.na(invite_2_outcome)) |> 
+    rename(from = invite_1_outcome, to = invite_2_outcome) |> 
+    summarise(
+      flow = n_distinct(participant_id, na.rm = T),
+      .by = c(from, to)
+    ),
+  
+  # flow from invite 1 to overall invite outcome (where invite 2 is NA)
+  df_milestones |> 
+    filter(!is.na(invite_1_outcome) & is.na(invite_2_outcome)) |> 
+    rename(from = invite_1_outcome, to = invite_outcome) |> 
+    summarise(
+      flow = n_distinct(participant_id, na.rm = T),
+      .by = c(from, to)
+    ),
+  
+  # flow from invite 2 to invite 3 (where not NA)
+  df_milestones |> 
+    filter(!is.na(invite_2_outcome) & !is.na(invite_3_outcome)) |> 
+    rename(from = invite_2_outcome, to = invite_3_outcome) |> 
+    summarise(
+      flow = n_distinct(participant_id, na.rm = T),
+      .by = c(from, to)
+    ),
+  
+  # flow from invite 2 to overall invite outcome (where invite 3 is NA)
+  df_milestones |> 
+    filter(!is.na(invite_2_outcome) & is.na(invite_3_outcome)) |> 
+    rename(from = invite_2_outcome, to = invite_outcome) |> 
+    summarise(
+      flow = n_distinct(participant_id, na.rm = T),
+      .by = c(from, to)
+    ),
+  
+  # final flow - invite 3 to overall outcome (where both are not NA)
+  df_milestones |> 
+    filter(!is.na(invite_3_outcome) & !is.na(invite_outcome)) |> 
+    rename(from = invite_3_outcome, to = invite_outcome) |> 
+    summarise(
+      flow = n_distinct(participant_id, na.rm = T),
+      .by = c(from, to)
+    )
+)
+
+# view our data
+df_flows |> 
+  reactable(defaultPageSize = 5)
+```
+
+# Sankey plot
+
+We now have a neat little summary of movements of participants between the milestones in our recruitment pathway. However, this 'tidy' data isn't the format required by [plotly](https://plotly.com/r/sankey-diagram/), so the next steps are to prepare it ready for plotting.
+
+## Preparing for plotly
+
+Plotly expects to be fed two sets of data, nodes and edges:
+
+1.  Nodes - these are the milestones we have in our `from` and `to` fields,
+
+2.  Edges - these are the flows that occur between nodes, the `flow` in our table.
+
+It is possible to extract this data by hand, however, I found using the [tidygraph](https://tidygraph.data-imaginist.com) package much easier and more convenient.
+
+```{r}
+df_sankey <- df_flows |> 
+  # convert our flows data to a tidy graph object
+  as_tbl_graph()
+
+```
+
+The tidygraph package splits our data into nodes and edges. We can selectively work on each by 'activating' them - here is the nodes list:
+
+```{r}
+df_sankey |> 
+  activate(what = 'nodes') |> 
+  as_tibble() |> 
+  reactable(defaultPageSize = 5)
+```
+
+You can see each unique node name listed. The row numbers for these nodes are used as reference ids in the edges object:
```

Comment:
Maybe 'ID'/'IDs' throughout.

Reply:
Thanks :) I've gone with IDs throughout the prose text